### PR TITLE
CZ2 - Initialize Database

### PR DESCRIPTION
**Pull request by malakivandas**

### CODE REVIEW

**Comment by malakivandas** [comment]
Created and tested the sql script.
Let me know what you think!

**Comment by malakivandas** [comment]
Done for now.

@BraidenCutforth please take a look and let me know what you think.

File: server/models/database_init.sh
Comment (BraidenCutforth) [inline server/models/database_init.sh]: You should add a newline to the end of your file. You can make this a setting in your editor.

File: server/models/database.sql
Comment (BraidenCutforth) [inline server/models/database.sql]: password length should be the size of the hash function you use. (sha256 => 256bits, 8 bytes) I am not sure if sha256 is still a suggested hash algo.

Comment (BraidenCutforth) [reply server/models/database.sql]: Also look up adding a salt.

File: server/models/database.sql
Comment (BraidenCutforth) [inline server/models/database.sql]: Put this in another sql file. Something relating to populating or dev data etc.

File: server/models/database.sql
Comment (BraidenCutforth) [inline server/models/database.sql]: Consider adding the `IF NOT EXISTS` specifiers

File: server/models/database.sql
Comment (BraidenCutforth) [inline server/models/database.sql]: What if a user wants to delete their account? Will we assign some specific ID?

File: server/models/database.sql
Comment (BraidenCutforth) [inline server/models/database.sql]: Since there are few categories you could make this bigger if you felt necessary.

**Review by BraidenCutforth — CHANGES_REQUESTED** [review]
Just a couple small things to consider.

Splitting the SQL into a create, delete and populate scripts. Making one shell script with flags will be nice.

File: server/models/database_init.sh
Comment (malakivandas) [reply server/models/database_init.sh]: Done. Thanks!

File: server/models/database.sql
Comment (malakivandas) [reply server/models/database.sql]: SHA-256 is hash function that outputs a value that is 256 bits long. That value can be represented most simply with either `CHAR(64)` (hex) or `BINARY(32)` ([byte string](https://qr.ae/pGB38P)). I will go with the byte string since it is [the optimized way](https://stackoverflow.com/a/16993150). (Makes sense.) 

The salt is added to the password before the hashing function so it will not affect the stored data type.

Since the authentication will be handled using AWS Cognito, the password will not be stored in the database. So, I did this as an exercise. Thanks for the suggestion!

File: server/models/database.sql
Comment (malakivandas) [reply server/models/database.sql]: Done.

File: server/models/database.sql
Comment (malakivandas) [reply server/models/database.sql]: CREATE TABLE will not overwrite an existing relation. IF NOT EXISTS only turns off the error message. I have decided that I want to see the error message.

File: server/models/database.sql
Comment (malakivandas) [reply server/models/database.sql]: 30 was the maximum number of characters from the list of categories that came from the Kamloops app. I will increase to 50 so it doesn't cause any problems with future insertions of categories.

File: server/models/database.sql
Comment (malakivandas) [reply server/models/database.sql]: I have added a 'deleted' flag to the users relation. Now, if a user deletes their account, their user record will not be deleted and their 'deleted' flag will be set to TRUE. The user field for an issue will be displayed as [deleted] (like reddit) if the user has been deleted.

Comment (malakivandas) [reply server/models/database.sql]: Good suggestion. Thank you!

Comment (malakivandas) [reply server/models/database.sql]: I'll leave this conversation unresolved and wait for your approval.

File: server/models/database.sql
Comment (malakivandas) [reply server/models/database.sql]: Actually, there is no `BINARY` keyword in Postgres so I will just use the other option, `CHAR(64)`.

Extra info:
- There exists a keyword BYTEA in Postgres, but the documentation is not very helpful and everyone just uses CHAR(64) with the hash stored in hex anyways.
- Postgres has an extension called pgcrypto that can handle the password salt, hashing, and storage. (I will not implement this since the account management will be handled by AWS Cognito.)

**Comment by malakivandas** [comment]
I've separated the SQL scripts into a create and a populate script and refactored the bash script with flags (--create, --populate, --delete, --reset).

Please review and let me know what you think! @BraidenCutforth

File: server/models/database.sh
Comment (BraidenCutforth) [inline server/models/database.sh:1]: This is wild haha. I am not equipped to review this to the full extent, I would just make sure that you follow a tutorial for writing a script with flags haha

File: server/models/database.sh
Comment (BraidenCutforth) [inline server/models/database.sh:18]: Nice!

**Review by BraidenCutforth — APPROVED** [review]

File: server/models/database.sh
Comment (malakivandas) [reply server/models/database.sh:1]: Why do you say wild? I did look at a video and a bunch of pages.

File: server/models/database.sql
Comment (BraidenCutforth) [reply server/models/database.sql]: This definitely meets the requirement but you aren't really deleting peoples information if they delete their account ahah. Another alternative is transfering the report to a special user in the database. You can do whatever you want.

File: server/models/database.sh
Comment (BraidenCutforth) [reply server/models/database.sh:1]: I say wild because the code is so hard to read but it is bash not you lol. In powershell you could do this:

```ps1
param(
    [switch]Create
    [switch]Delete
    ...
)

if($Delete) {
    # delete db command
}
# ... 
```

Notice that this reads a lot easier than the while true loop etc.

File: server/models/database.sql
Comment (malakivandas) [reply server/models/database.sql]: Yea I guess it will depend on how AWS Cognito will handle account deletion.

A lot of people said this is a good way to do it unless you wanted to have a deleted_users table, which only seemed to be useful for purging entries after a certain amount of time.

I'll leave this for now and address it when implementing AWS Cognito.

Thanks!

File: server/models/database.sh
Comment (malakivandas) [reply server/models/database.sh:1]: Yup that looks way better.
In bash you either use `case` like I did or use `getopts` which is better for handling arguments.
Going forward, maybe I'll put PowerShell on my Ubuntu then just write PowerShell scripts instead while still using zsh. I switched from using PowerShell in console because some things are unnecessarily harder than with Unix shells (e.g., setting the executable bit on a file `chmod +x <filename>`. Maybe I just wasn't able to find the easiest way.

Comment (malakivandas) [reply server/models/database.sh:1]: Thank you!

**Pull request merged**
